pkg/controller/seed-controller-manager/cluster-template-controller/controller.go

```diff
@@ -0,0 +1,312 @@
+	namespace string,
+	numWorkers int) error {
+
+	workerSelector, err := workerlabel.LabelSelector(workerName)
+	if err != nil {
+		return fmt.Errorf("failed to build worker-name selector: %v", err)
+	}
+
+	reconciler := &reconciler{
+		log:                     log.Named(ControllerName),
+		workerNameLabelSelector: workerSelector,
+		workerName:              workerName,
+		recorder:                mgr.GetEventRecorderFor(ControllerName),
+		namespace:               namespace,
+		seedClient:              mgr.GetClient(),
+	}
+
+	c, err := controller.New(ControllerName, mgr, controller.Options{Reconciler: reconciler, MaxConcurrentReconciles: numWorkers})
+	if err != nil {
+		return fmt.Errorf("failed to construct controller: %v", err)
+	}
+
+	if err := c.Watch(&source.Kind{Type: &kubermaticv1.ClusterTemplateInstance{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return fmt.Errorf("failed to create watch for seed cluster template instance: %v", err)
+	}
+
+	if err := c.Watch(
+		&source.Kind{Type: &kubermaticv1.Cluster{}},
+		enqueueClusterTemplateInstances(reconciler.seedClient, reconciler.log),
+		workerlabel.Predicates(workerName),
+	); err != nil {
+		return fmt.Errorf("failed to create watch for clusters: %w", err)
+	}
+
+	return nil
+}
+
+// Reconcile reconciles the kubermatic cluster template instance in the seed cluster
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+
+	log := r.log.With("request", request)
+	log.Debug("Reconciling")
+
+	instance := &kubermaticv1.ClusterTemplateInstance{}
+	if err := r.seedClient.Get(ctx, request.NamespacedName, instance); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to get cluster template instance %s: %w", instance.Name, ctrlruntimeclient.IgnoreNotFound(err))
+	}
+
+	err := r.reconcile(ctx, instance, log)
+	if err != nil {
+		log.Errorw("ReconcilingError", zap.Error(err))
+		r.recorder.Eventf(instance, corev1.EventTypeWarning, "ReconcilingError", err.Error())
+	}
+
+	return reconcile.Result{}, err
+
+}
+
+func (r *reconciler) reconcile(ctx context.Context, instance *kubermaticv1.ClusterTemplateInstance, log *zap.SugaredLogger) error {
+
+	clusterTemplateLabelSelector := ctrlruntimeclient.MatchingLabels{kubernetes.ClusterTemplateInstanceLabelKey: instance.Name}
+
+	clusterList := &kubermaticv1.ClusterList{}
+	if err := r.seedClient.List(ctx, clusterList, &ctrlruntimeclient.ListOptions{LabelSelector: r.workerNameLabelSelector}, clusterTemplateLabelSelector); err != nil {
+		return fmt.Errorf("failed listing clusters: %w", err)
+	}
+
+	// deletion
+	if !instance.DeletionTimestamp.IsZero() {
+		if !kuberneteshelper.HasFinalizer(instance, finalizer) {
+			return nil
+		}
+
+		if err := r.patchFinalizer(ctx, instance, true); err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	// initialization
+	if !kuberneteshelper.HasFinalizer(instance, finalizer) {
+		if err := r.patchFinalizer(ctx, instance, false); err != nil {
```

Comment:
NIT: could you add constants like 
```
var remove = true
var add = false
```
so when ```r.patchFinalizer(ctx, instance, remove)``` is called its more clear

Reply:
done